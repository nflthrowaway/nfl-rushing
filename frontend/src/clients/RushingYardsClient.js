import HttpClient from "./HttpClient";

class RushingYardsClient {
    constructor() {
        this.client = HttpClient;
    }

    async getRushingData(limit, skip, sort, sortDirection, filter) {
        const response = await this.client.get('', { params: { limit, skip, sort, sortDirection, filter }});
        return { results: response.data.results, total: response.data.total };
    }
}

export default RushingYardsClient;