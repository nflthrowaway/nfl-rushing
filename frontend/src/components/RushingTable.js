import React, { useEffect, useState } from 'react';

import RushingYardsClient from '../clients/RushingYardsClient';
import { columns, columnToMetricMap, metrics, metricToColumnMap } from '../common/constants';
import RushingTableHeaderRow from './RushingTableHeaderRow';
import RushingTablePagination from './RushingTablePagination';
import RushingTableRow from './RushingTableRow';
import RushingTableToolbar from './RushingTableToolbar';

const RushingTable = () => {
    const [loading, setLoading] = useState(true);
    const [rushingData, setRushingData] = useState([]);
    const [page, setPage] = useState(0);
    const [maxPage, setMaxPage] = useState(0);
    const [sortBy, setSortBy] = useState(columnToMetricMap['Player']);
    const [sortDirection, setSortDirection] = useState(1);
    const [filter, setFilter] = useState('');

    const paginationLimit = 20;
    const exportCsvPaginationLimit = 100;
    const itemDelimiter = ',';
    const rowDelimiter = '\n';

    const client = new RushingYardsClient();

    // Load data at start and when table changes occur (page, sort)
    useEffect(() => {
        client.getRushingData(paginationLimit, paginationLimit*page, sortBy, sortDirection, filter).then((data) => {
        setRushingData(data.results);
        setMaxPage(calculateMaxPage(data.total));
        setLoading(false);
        }).catch(err => console.log(err));
    }, [page, sortBy, sortDirection]);

    // Fetch data and reset page info when filtering
    const filterButtonOnClick = () => {
        client.getRushingData(paginationLimit, paginationLimit*page,  sortBy, sortDirection, filter).then((data) => {
        setRushingData(data.results);
        setPage(0);
        setMaxPage(calculateMaxPage(data.total));
        }).catch(err => console.log(err));
    };

    const convertRushingDataToCsvData = (rushingData, csvData) => {
        rushingData.forEach((item) => {
            // Reduce so we can combine LongestRush and LongestRushTD columns
            const reducedData = columns.reduce((result, column) => {
                if (column === metricToColumnMap.LongestRushTD) {
                    return result;
                }

                let itemData = item[columnToMetricMap[column]];
                if (column === metricToColumnMap.LongestRush && item['LongestRushTD']) {
                    itemData += 'T';
                }
                result.push(itemData);
                return result;
            }, []);
            csvData += reducedData.join(itemDelimiter) + rowDelimiter;
        });
        return csvData;
    }

    const downloadButtonOnClick = () => {
        let csvData = columns.filter(column => column !== metricToColumnMap.LongestRushTD).join(itemDelimiter) + rowDelimiter;
        let skip = 0;
        let rushingData;
        let total = 0;
        const promises = [];

        // First fetch to get total results at same time
        client.getRushingData(exportCsvPaginationLimit, skip, sortBy, sortDirection, filter).then((data) => {
            rushingData = data.results;
            total = data.total;
            csvData = convertRushingDataToCsvData(rushingData, csvData);
            skip += exportCsvPaginationLimit;
        
            // Get rest of pages by building promise array then executing
            while (skip < total) {
                promises.push(client.getRushingData(exportCsvPaginationLimit, skip, sortBy, sortDirection, filter).then((data) => {
                rushingData = data.results;
                csvData = convertRushingDataToCsvData(rushingData, csvData);
                }).catch(err => console.log(err)));
                skip += exportCsvPaginationLimit;
            }
        
            Promise.all(promises).then(() => {
                csvData = encodeURIComponent(csvData);
                csvData = 'data:text/csv;charset=utf-8,'+ csvData;
                const fileName = 'rushingData.csv';

                const downloadLink = document.createElement('a');
                downloadLink.setAttribute('href', csvData);
                downloadLink.setAttribute('download', fileName);
                downloadLink.click();
            }).catch(err => console.log(err));
        }).catch(err => console.log(err));
    };

    const calculateMaxPage = (totalResults) => {
        return Math.ceil(totalResults/paginationLimit);
    };

    const setSortByOnColumnClick = (metric) => {
        if (metrics.includes(metric)) {
            if (sortBy ===  metric) {
                setSortDirection(sortDirection*-1);
            }
            setSortBy(metric);
        }
    }

    return (
        <div className="TotalContainer">
            {loading ? 'Loading...' : (
                <div>
                    <RushingTableToolbar filterButtonOnClick={filterButtonOnClick} downloadButtonOnClick={downloadButtonOnClick} setFilter={setFilter}/>
                    <div className="TableContainer" data-testid={'table'}>
                        <div className='RushingTable'>
                            <table>
                                <colgroup>
                                    {columns.reduce((result, column, index) => {
                                        if (column !== metricToColumnMap.LongestRushTD) {
                                            result.push((
                                                <col key={index} className={`RushingTableColumn_${column}`}/>
                                            ));
                                        }
                                        return result;
                                    }, [])}
                                </colgroup>
                                <RushingTableHeaderRow setSort={setSortByOnColumnClick}/>
                                {rushingData.map((row, index) => (
                                    <RushingTableRow row={row} key={index}/>
                                ))}
                            </table>
                        </div>
                        <RushingTablePagination page={page} maxPage={maxPage} setPage={setPage}/>
                    </div>
                </div>
            )}
        </div>
    );
}

export default RushingTable;
