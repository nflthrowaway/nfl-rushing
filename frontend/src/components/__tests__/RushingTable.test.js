import { act, render } from '@testing-library/react';
import RushingTable from '../RushingTable';
import RushingYardsClient from '../../clients/RushingYardsClient';

jest.mock('../../clients/RushingYardsClient');

beforeEach(() => {
    RushingYardsClient.prototype.getRushingData = jest.fn().mockReturnValue(Promise.resolve({ results: [], total: 0 })); 
})

afterAll(() => {
    jest.restoreAllMocks();
});

test('renders loading text initially', () => {
    const { getByText } = render(<RushingTable />);
    const loadingText = getByText('Loading...');
    expect(loadingText).toBeInTheDocument();
});

test('renders empty table', async () => {
    const { findByTestId } = render(<RushingTable />);
    const table = await findByTestId('table');
    expect(table).toMatchSnapshot();
});

test('renders data in table', async () => {
    const testData = {
        Player: 'Name',
        Team: 'Team',
        Position: 'Position',
        RushingAttempts: 2,
        RushingAttemptsPerGame: 1,
        TotalRushingYards: 20,
        RushingAverageYardsPerAttempt: 10,
        RushingYardsPerGame: 20,
        TotalRushingTouchdowns: 0,
        LongestRush: 10,
        LongestRushTD: false,
        RushingFirstDowns: 0,
        RushingFirstDownPercentage: 0,
        Rushing20PlusYardsEach: 0,
        Rushing40PlusYardsEach: 0,
        RushingFumbles: 0
    };
    RushingYardsClient.prototype.getRushingData = jest.fn().mockReturnValue(Promise.resolve({ results: [testData], total: 0 })); 

    const { findByText, getByTestId } = render(<RushingTable />);
    await findByText('Name');
    const table = getByTestId('table');
    expect(table).toMatchSnapshot();
});
