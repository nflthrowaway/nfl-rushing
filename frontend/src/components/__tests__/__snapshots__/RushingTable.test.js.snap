// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders data in table 1`] = `
<div
  class="TableContainer"
  data-testid="table"
>
  <div
    class="RushingTable"
  >
    <table>
      <colgroup>
        <col
          class="RushingTableColumn_Player"
        />
        <col
          class="RushingTableColumn_Team"
        />
        <col
          class="RushingTableColumn_Pos"
        />
        <col
          class="RushingTableColumn_Att"
        />
        <col
          class="RushingTableColumn_Att/G"
        />
        <col
          class="RushingTableColumn_Yds"
        />
        <col
          class="RushingTableColumn_Avg"
        />
        <col
          class="RushingTableColumn_Yds/G"
        />
        <col
          class="RushingTableColumn_TD"
        />
        <col
          class="RushingTableColumn_Lng"
        />
        <col
          class="RushingTableColumn_1st"
        />
        <col
          class="RushingTableColumn_1st%"
        />
        <col
          class="RushingTableColumn_20+"
        />
        <col
          class="RushingTableColumn_40+"
        />
        <col
          class="RushingTableColumn_FUM"
        />
      </colgroup>
      <tbody>
        <tr
          class="RushingTableHeader"
        >
          <th>
            Player
          </th>
          <th>
            Team
          </th>
          <th>
            Pos
          </th>
          <th>
            Att
          </th>
          <th>
            Att/G
          </th>
          <th>
            Yds
          </th>
          <th>
            Avg
          </th>
          <th>
            Yds/G
          </th>
          <th>
            TD
          </th>
          <th>
            Lng
          </th>
          <th>
            1st
          </th>
          <th>
            1st%
          </th>
          <th>
            20+
          </th>
          <th>
            40+
          </th>
          <th>
            FUM
          </th>
        </tr>
      </tbody>
      <tbody>
        <tr
          class="RushingTableRow"
          data-testid="table-row"
        >
          <td
            class="RushingTableCell"
          >
            Name
          </td>
          <td
            class="RushingTableCell"
          >
            Team
          </td>
          <td
            class="RushingTableCell"
          >
            Position
          </td>
          <td
            class="RushingTableCell"
          >
            2
          </td>
          <td
            class="RushingTableCell"
          >
            1
          </td>
          <td
            class="RushingTableCell"
          >
            20
          </td>
          <td
            class="RushingTableCell"
          >
            10
          </td>
          <td
            class="RushingTableCell"
          >
            20
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
          <td
            class="RushingTableCell"
          >
            10
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
          <td
            class="RushingTableCell"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="PaginationButtons"
  >
    <button
      disabled=""
    >
      Prev
    </button>
    Page 1 of 0
    <button>
      Next
    </button>
  </div>
</div>
`;

exports[`renders empty table 1`] = `
<div
  class="TableContainer"
  data-testid="table"
>
  <div
    class="RushingTable"
  >
    <table>
      <colgroup>
        <col
          class="RushingTableColumn_Player"
        />
        <col
          class="RushingTableColumn_Team"
        />
        <col
          class="RushingTableColumn_Pos"
        />
        <col
          class="RushingTableColumn_Att"
        />
        <col
          class="RushingTableColumn_Att/G"
        />
        <col
          class="RushingTableColumn_Yds"
        />
        <col
          class="RushingTableColumn_Avg"
        />
        <col
          class="RushingTableColumn_Yds/G"
        />
        <col
          class="RushingTableColumn_TD"
        />
        <col
          class="RushingTableColumn_Lng"
        />
        <col
          class="RushingTableColumn_1st"
        />
        <col
          class="RushingTableColumn_1st%"
        />
        <col
          class="RushingTableColumn_20+"
        />
        <col
          class="RushingTableColumn_40+"
        />
        <col
          class="RushingTableColumn_FUM"
        />
      </colgroup>
      <tbody>
        <tr
          class="RushingTableHeader"
        >
          <th>
            Player
          </th>
          <th>
            Team
          </th>
          <th>
            Pos
          </th>
          <th>
            Att
          </th>
          <th>
            Att/G
          </th>
          <th>
            Yds
          </th>
          <th>
            Avg
          </th>
          <th>
            Yds/G
          </th>
          <th>
            TD
          </th>
          <th>
            Lng
          </th>
          <th>
            1st
          </th>
          <th>
            1st%
          </th>
          <th>
            20+
          </th>
          <th>
            40+
          </th>
          <th>
            FUM
          </th>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="PaginationButtons"
  >
    <button
      disabled=""
    >
      Prev
    </button>
    Page 1 of 0
    <button>
      Next
    </button>
  </div>
</div>
`;
